README.md
# flask-logservice

Log service for Flask.

## Installation

Start by downloading and building the project when necessary. The following
commands will do the job on most Debian based Linux distributions.

```bash
git clone https://github.com/marcbperez/flask-logservice
cd flask-logservice
export FLASK_APP="restfuloauth2"
export SECRET_KEY="non-production-key"
sudo -HE ./gradlew
```

## Usage

To start the service set the environment variables and run flask, it will be
available at `http://127.0.0.1:5000`.

```bash
sudo -HE flask run
```

First, access `http://127.0.0.1:5000` and create a user and api client. Then use
the client id, username and password to generate a bearer token.

```bash
curl -X POST -d \
"grant_type=password&client_id=8diLQbKSkseuZ99Q3kwFAWugXjDvImrqTALeM7sd\
&username=user&password=pass" \
http://127.0.0.1:5000/v1/oauth/token
```

The bearer token can now be used to access the user's protected data.

```bash
curl -H "Authorization: Bearer nOVFSNUDoP2bC1ScMRuYz8zCXeTY8F" \
http://127.0.0.1:5000/v1/oauth/check
```

The example log api is protected and will need a valid user, client and bearer
token. To list, add, modify and delete tasks see the script below.

```bash
# GET the log list.
curl -H "Authorization: Bearer yIMqTV5zOGQlRpIMBMpZnyHFMR0QW3" \
http://127.0.0.1:5000/v1/log
# Paginate and sort de log list by ascending (asc) or descending (desc).
curl -H "Authorization: Bearer yIMqTV5zOGQlRpIMBMpZnyHFMR0QW3" \
"http://127.0.0.1:5000/v1/log?page=1&max_results=5&sort=id-desc"
# CREATE a new log.
curl -X POST -H "Authorization: Bearer yIMqTV5zOGQlRpIMBMpZnyHFMR0QW3" \
-d "public=0&message=Hello world!" \
http://127.0.0.1:5000/v1/log
# MODIFY a log stating its etag.
curl -X PUT -H "Authorization: Bearer yIMqTV5zOGQlRpIMBMpZnyHFMR0QW3" \
-d "etag=rz05FPx8qOIYJdmYZNLvcWupzh9qLlSoZnphpBFC\
    &public=1&message=Hello universe!" \
http://127.0.0.1:5000/v1/log/1
# GET a log.
curl -H "Authorization: Bearer yIMqTV5zOGQlRpIMBMpZnyHFMR0QW3" \
http://127.0.0.1:5000/v1/log/1
# DELETE a log stating its etag.
curl -X DELETE -H "Authorization: Bearer yIMqTV5zOGQlRpIMBMpZnyHFMR0QW3" \
-d "etag=DMwVytdmg3CtwDgbm9wWOINjX73Iev2n4NFkRsV7" \
http://127.0.0.1:5000/v1/log/1
```

A search operation is also available. To use it send an url-encoded JSON
`search` parameter to an item endpoint such as `http://127.0.0.1:5000/v1/log`.
Valid column condition operators are `and` and `or`. As for column operators all
 `=`, `!=`, `<`, `<=`, `>`, `>=` and `like` are available.

```json
{
  "operator": "and",
  "conditions": [
    {
      "column": "id",
      "operator": "=",
      "value": 3
    },
    {
      "operator": "or",
      "conditions": [
        {
          "column": "public",
          "operator": "=",
          "value": 1
        },
        {
          "column": "etag",
          "operator": "!=",
          "value": ""
        }
      ]
    }
  ]
}
```

## Testing

Test checks are executed automatically every time the project is built. Builds
can be done remotely or continuously on a development context. For continuous
integration and development use docker-compose. This is recommended to keep the
system clean while the project is built every time the sources change.

```bash
sudo docker-compose up
```

For continuous integration and development without any dependencies use the
Gradle wrapper. This is the best option if the wrapper is available and the
Docker context is not valid. For a full list of tasks, see
`sudo ./gradlew tasks --all`. For a CI cycle use `sudo ./gradlew --continuous`.

For continuous integration and development without Docker or the project wrapper
use Gradle directly. This will create the wrapper in case it is not present.
Similar to the above, for a CI cycle use `sudo gradle --continuous`. Gradle
3.4.1 is required for this to work. Plain Docker is also available for remote
integration tasks and alike. Build the image with `sudo docker build .` and run
a new container with it. Information on how to install Docker and docker-compose
can be found in their [official page][install-docker-compose]. A similar
installation guide is available [for Gradle][install-gradle].

## Troubleshooting

The [issue tracker][issue-tracker] intends to manage and compile bugs,
enhancements, proposals and tasks. Reading through its material or reporting to
its contributors via the platform is strongly recommended.

## Contributing

This project adheres to [Semantic Versioning][semver] and to certain syntax
conventions defined in [.editorconfig][editorconfig]. To get a list of changes
refer to the [CHANGELOG][changelog]. Only branches prefixed by *feature-*,
*hotfix-*, or *release-* will be considered:

  - Fork the project.
  - Create your new branch: `git checkout -b feature-my-feature develop`
  - Commit your changes: `git commit -am 'Added my new feature.'`
  - Push the branch: `git push origin feature-my-feature`
  - Submit a pull request.

## Credits

This project is created by [marcbperez][author] and maintained by its
[author][author] and contributors.

## License

This project is licensed under the [Apache License Version 2.0][license].

[author]: https://marcbperez.github.io
[issue-tracker]: https://github.com/marcbperez/flask-logservice/issues
[editorconfig]: .editorconfig
[changelog]: CHANGELOG.md
[license]: LICENSE
[semver]: http://semver.org
[install-docker-compose]: https://docs.docker.com/compose/install/
[install-gradle]: https://gradle.org/install


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This
project adheres to [Semantic Versioning](http://semver.org).

## 0.3.1 - 2017-05-07

### Fixed

  - Gradle wrapper for this project.
  - Wrong information in README files.

## 0.3.0 - 2017-05-07

### Added

  - Test cases for python module.
  - Log service module, database and base files.
  - Log service model and endpoint.
  - Static files for this module.

### Fixed

  - Wrong parameters in module setup script.

## 0.2.0 - 2017-05-06

### Added

  - Gradle build script for this project.
  - Gradle wrapper for this project.
  - Dockerfile and docker-compose manifest for this project.
  - Module manifest, configuration and setup scripts.
  - README in reStructuredText format.
  - Cases for the build environment.
  - Cases for the supporting files.

## 0.1.0 - 2017-05-05

### Added

  - A README which covers project description and usage.
  - This CHANGELOG file to document project changes and versions.
  - An EditorConfig to suggest syntax and format standards.
  - A LICENSE file to define project ownership and rights.
  - A .gitignore list for untracked resources.


tests/test_restfuloauth2.py
import pytest
import restfuloauth2
from restfuloauth2.database import db
from restfuloauth2.oauth.models import Client
import json
import urllib


@pytest.fixture
def client(request):
    """Configures the application and returns the test client."""
    restfuloauth2.app.config['TESTING'] = True
    restfuloauth2.app.config['SECRET_KEY'] = 'testing-key'

    db.drop_all(app=restfuloauth2.app)
    db.create_all(app=restfuloauth2.app)

    return restfuloauth2.app.test_client()


def create_user(client, username, password):
    """Creates a new user given a username and password."""
    rv = client.post('/v1/oauth/', follow_redirects=True, data={
        'submit': 'Add User',
        'username': username,
        'password': password,
    })


def create_client(client):
    """Creates a new client."""
    rv = client.post('/v1/oauth/', follow_redirects=True, data={
        'submit': 'Add Client',
    })

    db.app = restfuloauth2.app
    oauth_clients = Client.query.all()
    client_id = oauth_clients[0].client_id

    return client_id


def create_token(client, username, password, client_id):
    """Creates a new token given a username, password and client id."""
    rv = client.post('/v1/oauth/token', follow_redirects=True, data={
        'grant_type': 'password',
        'client_id': client_id,
        'username': username,
        'password': password
    })

    json_response = json.loads(rv.data)
    access_token = json_response['access_token']

    return access_token


def test_default_redirect(client):
    """Expects the default route to redirect to management."""
    rv = client.get('/', follow_redirects=True)
    assert 'Users' in rv.data
    assert 'Clients' in rv.data


def test_check(client):
    """Expects to obtain the authenticated user greeting."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    rv = client.get('/v1/oauth/check', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    })

    assert 'userA' in rv.data


def test_get_log_index(client):
    """Expects to get and empty index."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    rv = client.get('/v1/log', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    })

    assert '[]' in rv.data


def test_get_nonexistent_log(client):
    """Expects a not found error from a nonexistent item."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    rv = client.get('/v1/log/1', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    })

    assert rv.status_code == 404


def test_create_log(client):
    """Expects to create a new item."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    # Create the item.
    rv = client.post('/v1/log', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    }, data={
        'public': 0,
        'message': 'Hello world!',
    })

    assert rv.status_code == 201

    # Retrieve the new item.
    rv = client.get('/v1/log/1', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token})

    assert rv.status_code == 200


def test_update_log(client):
    """Expects to update an existing item."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    # Create the item.
    rv = client.post('/v1/log', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    }, data={
        'public': 0,
        'message': 'Hello world!',
    })

    # Get the item's etag.
    json_response = json.loads(rv.data)
    etag = json_response['etag']

    # Try a wrong etag and expect an error.
    rv = client.put('/v1/log/1', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    }, data={
        'etag': etag + 'a',
        'public': 1,
        'message': 'Hello universe!',
    })

    assert rv.status_code != 201

    # Try with the right etag.
    rv = client.put('/v1/log/1', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    }, data={
        'etag': etag,
        'public': 1,
        'message': 'Hello universe!',
    })

    assert rv.status_code == 201


def test_delete_log(client):
    """Expects to delete an existing item."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    # Create the item.
    rv = client.post('/v1/log', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    }, data={
        'public': 0,
        'message': 'Hello world!',
    })

    # Get the item's etag.
    json_response = json.loads(rv.data)
    etag = json_response['etag']

    # Try a wrong etag and expect an error.
    rv = client.delete('/v1/log/1', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    }, data={
        'etag': etag + 'a',
    })

    assert rv.status_code != 204

    # Try with the right etag.
    rv = client.delete('/v1/log/1', follow_redirects=True, headers={
        'Authorization': 'Bearer ' + access_token
    }, data={
        'etag': etag,
    })

    assert rv.status_code == 204


def test_pagination(client):
    """Expects to get a paginated collection."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    # Create 20 new items.
    for i in range(20):
        rv = client.post('/v1/log', follow_redirects=True, headers={
            'Authorization': 'Bearer ' + access_token
        }, data={
            'public': 0,
            'message': 'Hello world!',
        })

    # Get the paginated collection.
    rv = client.get(
        '/v1/log?page=2&max_results=5&sort=id-desc', follow_redirects=True,
        headers={'Authorization': 'Bearer ' + access_token})

    json_response = json.loads(rv.data)
    assert json_response[0]['id'] is 15
    assert len(json_response) is 5


def test_search(client):
    """Expects to search a collection."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    # Create 5 new items.
    for i in range(5):
        rv = client.post('/v1/log', follow_redirects=True, headers={
            'Authorization': 'Bearer ' + access_token
        }, data={
            'public': 0,
            'message': 'Hello world!',
        })

    # Make a simple search by id.
    search = urllib.quote('{"column": "id","operator": "=","value": 3}')
    rv = client.get(
        '/v1/log?search=' + search, follow_redirects=True,
        headers={'Authorization': 'Bearer ' + access_token})

    json_response = json.loads(rv.data)
    assert json_response[0]['id'] is 3
    assert len(json_response) is 1

    # Try a wrong column operator and expect an error.
    wrong_column_operator = urllib.quote(
        '{"column": "id","operator": "x","value": 3}')
    rv = client.get(
        '/v1/log?search=' + wrong_column_operator, follow_redirects=True,
        headers={'Authorization': 'Bearer ' + access_token})

    assert rv.status_code == 401

    # Try a string value.
    search_text = urllib.quote(
        '{"column": "etag","operator": "!=","value": ""}')
    rv = client.get(
        '/v1/log?search=' + search_text, follow_redirects=True,
        headers={'Authorization': 'Bearer ' + access_token})

    json_response = json.loads(rv.data)
    assert rv.status_code == 200
    assert len(json_response) is 5


def test_complex_search(client):
    """Expects to do a complex search on a collection."""
    create_user(client, 'userA', 'passA')
    client_id = create_client(client)
    access_token = create_token(client, 'userA', 'passA', client_id)

    search = urllib.quote('{\
      "operator": "and",\
      "conditions": [\
        {\
          "column": "id",\
          "operator": "=",\
          "value": 3\
        },\
        {\
          "operator": "or",\
          "conditions": [\
            {\
              "column": "public",\
              "operator": "=",\
              "value": 0\
            },\
            {\
              "column": "etag",\
              "operator": "!=",\
              "value": ""\
            }\
          ]\
        }\
      ]\
    }')

    wrong_condition_operator = urllib.quote('{\
      "operator": "x",\
      "conditions": [\
        {\
          "column": "id",\
          "operator": "=",\
          "value": 3\
        },\
        {\
          "column": "etag",\
          "operator": "!=",\
          "value": ""\
        }\
      ]\
    }')

    # Create 5 new items.
    for i in range(5):
        rv = client.post('/v1/log', follow_redirects=True, headers={
            'Authorization': 'Bearer ' + access_token
        }, data={
            'public': 0,
            'message': 'Hello world!'
        })

    # Make a complex search configured with a JSON object.
    rv = client.get(
        '/v1/log?search=' + search, follow_redirects=True,
        headers={'Authorization': 'Bearer ' + access_token})

    json_response = json.loads(rv.data)
    assert json_response[0]['id'] is 3
    assert len(json_response) is 1

    # Try a wrong condition operator and expect an error.
    rv = client.get(
        '/v1/log?search=' + wrong_condition_operator, follow_redirects=True,
        headers={'Authorization': 'Bearer ' + access_token})

    assert rv.status_code == 401


restfuloauth2/log/__init__.py
from ..database import db
from ..database.model import Model
from werkzeug.security import gen_salt
from datetime import datetime


class Log(db.Model, Model):
    """Minimal implementation of a REST model."""

    __tablename__ = 'log'

    message = db.Column(db.String(1000))

    def serialize(self):
        """Returns the serialized version of this model."""
        serialized = Model.serialize(self)
        serialized['message'] = self.message
        return serialized

    @classmethod
    def add_parser_args(cls, parser):
        """Adds model parameters to parser validation."""
        parser.add_argument('public', required=True, help='Model visibility.')
        parser.add_argument('message', required=True, help='Model message.')

    @classmethod
    def update(cls, model, etag, public, message):
        """Updates a model if the etag matches."""
        if model.etag != etag:
            return False

        model.etag = gen_salt(40)
        model.updated = datetime.utcnow()
        model.public = public == '1'
        model.message = message
        db.session.commit()

        return model

    @classmethod
    def create(cls, user, public, message):
        """Creates a model owned by the provided user."""
        model = cls()
        model.user_id = user.id
        model.etag = gen_salt(40)
        model.created = model.updated = datetime.utcnow()
        model.public = public == '1'
        model.message = message
        db.session.add(model)
        db.session.commit()

        return model


restfuloauth2/log/endpoint.py
from . import Log
from ..oauth.models import User
from ..database.query import Query
from flask import request
from flask_restful import abort, Resource


class LogItem(Resource):
    """Log model item endpoint."""

    def get(self, log_id):
        """Gets a model given its id and outputs the serialized version."""
        user = User.get_authorized()
        log = Query.get_item_or_abort(Log, log_id, user)

        return log.serialize()

    def delete(self, log_id):
        """Deletes a model given its id."""
        user = User.get_authorized()
        log = Query.get_item_or_abort(Log, log_id, user)
        args = Log.parse_delete_arguments()
        delete = Log.delete(log, args['etag'])

        if not delete:
            abort(401, message=Query.ETAG_NOT_MATCHING)

        return '', 204

    def put(self, log_id):
        """Updates a model given its id and outputs the serialized version."""
        user = User.get_authorized()
        log = Query.get_item_or_abort(Log, log_id, user)
        args = Log.parse_put_arguments()
        update = Log.update(log, args['etag'], args['public'], args['message'])

        if not update:
            abort(401, message=Query.ETAG_NOT_MATCHING)

        return update.serialize(), 201


class LogIndex(Resource):
    """Log model index endpoint."""

    def get(self):
        """Outputs a serialized, paginated collection of models."""
        page = request.args.get('page', Query.DEFAULT_PAGE)
        max_results = request.args.get(
            'max_results', Query.DEFAULT_MAX_RESULTS)
        sort = request.args.get('sort', Query.DEFAULT_SORT)
        sort_direction = Query.get_sort_attribute(Log, sort)
        search = request.args.get('search', Query.DEFAULT_SEARCH)

        user = User.get_authorized()
        logs = Log.get_permitted_models(
            user, sort_direction, page, max_results, search)

        return Log.serialize_list(logs)

    def post(self):
        """Adds a model to the colletion and outputs the serialized version."""
        user = User.get_authorized()
        args = Log.parse_post_arguments()
        log = Log.create(user, args['public'], args['message'])

        return log.serialize(), 201
